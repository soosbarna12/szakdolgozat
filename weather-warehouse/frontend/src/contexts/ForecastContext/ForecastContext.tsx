import React, { createContext, useEffect, useState } from "react";
import { ForecastContextType, ForecastLocationData, ForecastProviderProps } from "./ForecastContext.type";

export const ForecastContext = createContext<ForecastContextType>({
  location: { name: "", lat: 0, lon: 0 },
  setLocation: () => {},
});

export const ForecastLocationProvider = ({ children }: ForecastProviderProps) => {
  const [location, setLocation] = useState<ForecastLocationData>(() => {
    const storedLocation = localStorage.getItem("forecastLocation");
    const defaultLocation = { name: "", lat: 0, lon: 0 };
    try {
      return storedLocation ? JSON.parse(storedLocation) : defaultLocation;
    } catch (error) {
      return defaultLocation;
    }
  });

  useEffect(() => {
    // Persist location to localStorage whenever it changes
    localStorage.setItem("forecastLocation", JSON.stringify(location));
  }, [location]);

  return (
    <ForecastContext.Provider value={{ location, setLocation }}>
      {children}
    </ForecastContext.Provider>
  );
};