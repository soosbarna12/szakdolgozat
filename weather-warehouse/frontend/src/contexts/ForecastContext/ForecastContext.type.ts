import { Dispatch, ReactNode, SetStateAction } from "react";

export interface ForecastLocationData {
  name: string;
  country?: string;
  lat: number;
  lon: number;
}

export interface ForecastContextType {
  location: ForecastLocationData;
  setLocation: Dispatch<SetStateAction<ForecastLocationData>>;
}

export interface ForecastProviderProps {
  children: ReactNode;
}