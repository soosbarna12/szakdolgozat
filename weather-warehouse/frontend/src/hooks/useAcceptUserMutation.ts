import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";


export function useAcceptUserMutation() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (id: number) => {
      const response = await axios.post(`/user/accept`, { id });
      return response.data;
    },
    onSuccess: () => {
      showAlert("User accepted successfully", "success");
    },
    onError: () => {
      showAlert("Failed to accept user", "error");
    },
  });

  return mutation;
}
