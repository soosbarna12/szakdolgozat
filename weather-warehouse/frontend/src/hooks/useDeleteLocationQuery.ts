import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";

export function useDeleteLocationQuery() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (userLocationID: number) => {
      const response = await axios.delete(`/user/deleteLocation`, { params: { userLocationID } });
      return response.data;
    },
    onSuccess: () => {
      showAlert("Location deleted successfully", "success");
    },
    onError: () => {
      showAlert("Failed to delete location", "error");
    },
  });

  return mutation;
}