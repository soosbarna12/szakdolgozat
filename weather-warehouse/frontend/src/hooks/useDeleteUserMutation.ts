import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";


export function useDeleteUserMutation() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (id: number) => {
      const response = await axios.delete(`/user/delete`, { params: { id } });
      return response.data;
    },
    onSuccess: () => {
      showAlert('User deleted successfully', 'success');
    },
    onError: () => {
      showAlert('Failed to delete user', 'error');
    },
  });

  return mutation;
}
