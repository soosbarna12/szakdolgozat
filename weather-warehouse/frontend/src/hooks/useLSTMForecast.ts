import { useQuery } from "@tanstack/react-query";
import axios from "../utils/axiosConfig";

export function useLSTMForecast(locationName: string) {
  return useQuery({
    queryKey: ["forecast", locationName],
    queryFn: async () => {
      const response = await axios.post("/api/forecast/forecastLSTM", { location: locationName });
      return response.data.forecast;
    },
    enabled: !!locationName,
  });
}