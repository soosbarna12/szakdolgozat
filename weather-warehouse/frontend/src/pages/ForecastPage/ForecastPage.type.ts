export interface ForecastPageProps {
	dayName: string;
}
