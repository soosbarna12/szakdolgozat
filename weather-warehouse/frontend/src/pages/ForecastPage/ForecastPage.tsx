import React, { useContext } from "react";
import Grid from "@mui/material/Grid2";
import { FilterBar } from "../../components/FilterBar/FilterBar";
import { ForecastContext } from "../../contexts/ForecastContext/ForecastContext";
import { useLSTMForecast } from "../../hooks/useLSTMForecast";
import { Pages } from "../../types/page.type";
import { ContentBox } from "../../stlyes/content.style";
import { ForecastTile } from "../../components/DataGrids/ForecastTileGroup/ForecastTile/ForecastTile";
import { Skeleton } from "@mui/material";

export function ForecastPage() {
  const { location } = useContext(ForecastContext);
  const { data: forecastData, isLoading, error } = useLSTMForecast(location.name);

  function renderSkeletons() {
    if (!isLoading) return null;
    return [0, 1, 2, 3, 4, 5, 6].map((index) => (
      <Grid key={index}>
        <Skeleton
          data-testid="forecastTileSkeleton"
          variant="rectangular"
          animation="wave"
          width="180px"
          height="150px"
          sx={{ borderRadius: '10px' }}
        />
      </Grid>
    ));
  }

  function renderError() {
    if (!error) return null;
    return <p>Error fetching forecast data.</p>;
  }

  function renderForecastTiles() {
    if (isLoading || error) return null;
    return forecastData?.map((day: { date: string; temperature: number | null }, index: number) => (
      <Grid key={index} sx={{ paddingBottom: "2px" }}>
        <ForecastTile
          date={day.date}
          temperature={day.temperature != null ? day.temperature.toFixed(2) : "N/A"}
        />
      </Grid>
    ));
  }

  return (
    <>
      <FilterBar type={Pages.Forecast} location={location.name} />
      <ContentBox sx={{ display: "flex", justifyContent: "center", paddingBottom: "2px" }}>
        <Grid container spacing={2} justifyContent="center" alignItems="center">
          {renderSkeletons()}
          {renderError()}
          {renderForecastTiles()}
        </Grid>
      </ContentBox>
    </>
  );
}