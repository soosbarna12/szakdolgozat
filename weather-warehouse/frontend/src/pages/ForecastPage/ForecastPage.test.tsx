import '@testing-library/jest-dom';
import { screen, waitFor } from '@testing-library/react';
import React from 'react';
import { ForecastPage } from './ForecastPage';
import { ForecastContext } from '../../contexts/ForecastContext/ForecastContext';
import { useLSTMForecast } from '../../hooks/useLSTMForecast';
import { renderWithQueryClient } from '../../utils/test/renderWithQueryClient';

jest.mock('../../hooks/useLSTMForecast', () => ({
  useLSTMForecast: jest.fn(),
}));

jest.mock('react-leaflet', () => ({
  MapContainer: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="dataMapContainer">{children}</div>
  ),
  TileLayer: () => <div data-testid="mockTileLayer">Mock TileLayer</div>,
  useMap: jest.fn(() => ({
    setView: jest.fn(),
  })),
}));

describe('pages/ForecastPage', () => {
  const mockLocation = { name: 'Budapest', lat: 47.4979, lon: 19.0402 };

  const renderForecastPage = () =>
    renderWithQueryClient(
      <ForecastContext.Provider value={{ location: mockLocation, setLocation: jest.fn() }}>
        <ForecastPage />
      </ForecastContext.Provider>
    );

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches ForecastPage component snapshot', () => {
    (useLSTMForecast as jest.Mock).mockReturnValue({
      data: null,
      error: null,
      isLoading: true,
    });

    const { container } = renderForecastPage();
    expect(container).toMatchSnapshot();
  });

  it('renders skeletons when loading', () => {
    (useLSTMForecast as jest.Mock).mockReturnValue({
      data: null,
      error: null,
      isLoading: true,
    });

    renderForecastPage();

    const skeletons = screen.getAllByTestId('forecastTileSkeleton');
    expect(skeletons.length).toBeGreaterThan(0);
  });

  it('renders forecast tiles when data is available', async () => {
    const mockForecastData = [
      { date: '2025-05-08', temperature: 22.5 },
      { date: '2025-05-09', temperature: 24.0 },
    ];

    (useLSTMForecast as jest.Mock).mockReturnValue({
      data: mockForecastData,
      error: null,
      isLoading: false,
    });

    renderForecastPage();

    await waitFor(() => {
      const forecastTiles = screen.getAllByTestId('forecastTile');
      expect(forecastTiles.length).toBe(mockForecastData.length);
    });
  });
});