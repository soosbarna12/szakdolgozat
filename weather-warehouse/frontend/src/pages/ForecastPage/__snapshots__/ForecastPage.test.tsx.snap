// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/ForecastPage matches ForecastPage component snapshot 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionStatic css-94iixj-MuiPaper-root-MuiAppBar-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-t4zkma-MuiToolbar-root"
    >
      <div
        class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1tlcqt-MuiAutocomplete-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-rs1at4-MuiFormControl-root-MuiTextField-root"
          data-testid="locationSearchBox"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1e3qgpi-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1nowbqt-MuiInputAdornment-root"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="SearchOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
            </div>
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-qve9l1-MuiInputBase-input-MuiOutlinedInput-input"
              id="«r0»"
              placeholder="Location"
              role="combobox"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment css-1uhhrmm-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-2y136s-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-w4cd9x"
              >
                <span
                  aria-hidden="true"
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    class="MuiBox-root css-1l5d7sl"
  >
    <div
      class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row MuiGrid2-spacing-xs-2 css-1yfq57-MuiGrid2-root"
    >
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row css-7yij0q-MuiGrid2-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-ctxb8s-MuiPaper-root"
          data-testid="forecastTileSkeleton"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-wave css-1km0t8c-MuiSkeleton-root"
            style="width: 100%; height: 100%;"
          />
        </div>
      </div>
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row css-7yij0q-MuiGrid2-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-ctxb8s-MuiPaper-root"
          data-testid="forecastTileSkeleton"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-wave css-1km0t8c-MuiSkeleton-root"
            style="width: 100%; height: 100%;"
          />
        </div>
      </div>
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row css-7yij0q-MuiGrid2-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-ctxb8s-MuiPaper-root"
          data-testid="forecastTileSkeleton"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-wave css-1km0t8c-MuiSkeleton-root"
            style="width: 100%; height: 100%;"
          />
        </div>
      </div>
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row css-7yij0q-MuiGrid2-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-ctxb8s-MuiPaper-root"
          data-testid="forecastTileSkeleton"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-wave css-1km0t8c-MuiSkeleton-root"
            style="width: 100%; height: 100%;"
          />
        </div>
      </div>
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row css-7yij0q-MuiGrid2-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-ctxb8s-MuiPaper-root"
          data-testid="forecastTileSkeleton"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-wave css-1km0t8c-MuiSkeleton-root"
            style="width: 100%; height: 100%;"
          />
        </div>
      </div>
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row css-7yij0q-MuiGrid2-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-ctxb8s-MuiPaper-root"
          data-testid="forecastTileSkeleton"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-wave css-1km0t8c-MuiSkeleton-root"
            style="width: 100%; height: 100%;"
          />
        </div>
      </div>
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row css-7yij0q-MuiGrid2-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-ctxb8s-MuiPaper-root"
          data-testid="forecastTileSkeleton"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-wave css-1km0t8c-MuiSkeleton-root"
            style="width: 100%; height: 100%;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
